server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -112,20 +103,59 @@ void sendErrorResponse(final Version nodeVersion, final TcpChannel channel, fina
     }
 
     private void sendMessage(TcpChannel channel, OutboundMessage networkMessage, ActionListener<Void> listener) throws IOException {
-        MessageSerializer serializer = new MessageSerializer(networkMessage, bigArrays);
-        SendContext sendContext = new SendContext(channel, serializer, listener, serializer);
-        internalSend(channel, sendContext);
+        final BytesStreamOutput bytesStreamOutput = new ReleasableBytesStreamOutput(bigArrays);
+        final ActionListener<Void> wrappedListener = ActionListener.runBefore(listener, bytesStreamOutput::close);
+        final BytesReference message;
+        try {
+            message = networkMessage.serialize(bytesStreamOutput);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("failed to serialize outbound message [{}]", networkMessage), e);
+            wrappedListener.onFailure(e);
+            throw e;
+        }
+        internalSend(channel, message, networkMessage, wrappedListener);
     }
 
-    private void internalSend(TcpChannel channel, SendContext sendContext) throws IOException {
-        channel.getChannelStats().markAccessed(threadPool.relativeTimeInMillis());
-        BytesReference reference = sendContext.get();
+    private void internalSend(TcpChannel channel, BytesReference reference, @Nullable OutboundMessage message,
+                              ActionListener<Void> listener) {
+        final long startTime = threadPool.relativeTimeInMillis();
```

Comment:
This also apparently fixes a small bug, see https://discuss.elastic.co/t/org-elasticsearch-action-get-multigetshardresponse-extremly-high-took-value/294434.